import_export/tmp_storages.py

```diff
@@ -63,7 +63,10 @@ def save(self, data, mode=None):
         cache.set(self.CACHE_PREFIX + self.name, data, self.CACHE_LIFETIME)
 
     def read(self, read_mode='r', encoding=None):
-        return cache.get(self.CACHE_PREFIX + self.name)
+        data = cache.get(self.CACHE_PREFIX + self.name)
+        if isinstance(data, bytes) and encoding is not None:
+            data = data.decode(encoding)
```

Comment:
@matthewhegarty for some reason I don't need to use `force_str`, decode works well here.

Reply:
I don't mind doing it this way but it would be ideal to document that the return type is a string if encoding is provided otherwise is a byte array.

It seems like a code smell that you get a different return type depending on the encoding flag.  It probably should always return a string given that clients supply a string when calling `save()`, and also that the other `tmp_storages` implementations return strings.  However if we always return a string (not bytearray) could potentially break things for existing implementations, so if we did it that way we'd have to mark it as a v3.0 PR.  I would favour doing it this way.l

What are your thoughts?

Reply:
The idea is that you supply the encoding when the data is not binary (e.g. CSV). I added a test for CacheStorage and binary data using books.xls. Maybe we need more tests to be sure we get this issue right.

[Your PR](https://github.com/django-import-export/django-import-export/pull/1306) is already marked with "breaking change", so no problem.

What I didn't like was the magic decoding in `read_from_tmp_storage`, like the tmp_storages classes are not taking full responsibility, or that the interface is not clear.

This was supposed to go into your PR, but I was lazy to create the PR against your repo, so I pushed your branch here.